modules/calib3d/src/chessboard.cpp

```diff
@@ -0,0 +1,3295 @@
+            return false;
+    }
+    std::vector<cv::Point2f>::const_iterator iter = points.begin();
+    std::vector<cv::Point2f>::const_iterator iter2 = iter+1;
+    std::vector<cv::Point2f>::const_iterator iter3 = iter2+1;
+    std::vector<cv::Point2f>::const_iterator iter4 = iter3+1;
+    Ellipse ellipse;
+    if(!estimateSearchArea(*iter4,*iter3,*iter2,CORNERS_SEARCH*3,ellipse))
+        return false;
+    if(!ellipse.contains(*iter))
+        return false;
+
+    std::vector<cv::Point2f>::const_iterator iter5 = iter4+1;
+    for(;iter5 != points.end();++iter5)
+    {
+        if(!estimateSearchArea(*iter2,*iter3,*iter4,CORNERS_SEARCH,ellipse,&(*iter)))
+            return false;
+        if(!ellipse.contains(*iter5))
+            return false;
+        iter = iter2;
+        iter2 = iter3;
+        iter3 = iter4;
+        iter4 = iter5;
+    }
+    return true;
+}
+
+cv::Point2f &Chessboard::Board::getCorner(int _row,int _col)
+{
+    int _rows = int(rowCount());
+    int _cols = int(colCount());
+    if(_row >= _rows || _col >= _cols)
+        CV_Error(Error::StsBadArg,"out of bound");
+    if(_row == 0)
+    {
+        PointIter iter(top_left,TOP_LEFT);
+        int count = 0;
+        do
+        {
+            if(count == _col)
+                return *(*iter);
+            ++count;
+        }while(iter.right());
+    }
+    else
+    {
+        Cell *row_start = top_left;
+        int count = 1;
+        do
+        {
+            if(count == _row)
+            {
+                PointIter iter(row_start,BOTTOM_LEFT);
+                int count2 = 0;
+                do
+                {
+                    if(count2 == _col)
+                        return *(*iter);
+                    ++count2;
+                }while(iter.right());
+            }
+            ++count;
+            row_start = row_start->bottom;
+        }while(_row);
+    }
+    CV_Error(Error::StsInternal,"cannot find corner");
+}
+
+bool Chessboard::Board::isCellBlack(int row,int col)const
+{
+    return getCell(row,col)->black;
+}
+
+bool Chessboard::Board::isCellEmpty(int row,int col)
+{
+    return getCell(row,col)->empty();
+}
+
+Chessboard::Board::Cell* Chessboard::Board::getCell(int row,int col)
+{
+    const Cell *cell = const_cast<const Board*>(this)->getCell(row,col);
+    return const_cast<Cell*>(cell);
+}
+
+const Chessboard::Board::Cell* Chessboard::Board::getCell(int row,int col)const
+{
+    if(row > rows-1 || row < 0 || col > cols-1 || col < 0)
+        CV_Error(Error::StsBadArg,"out of bound");
+    PointIter p_iter(top_left,BOTTOM_RIGHT);
+    for(int i=0; i< row; p_iter.bottom(),++i);
+    for(int i=0; i< col; p_iter.right(),++i);
+    return p_iter.getCell();
+}
+
+
+bool Chessboard::Board::isEmpty()const
+{
+    return cells.empty();
+}
+
+size_t Chessboard::Board::colCount()const
+{
+    return cols;
+}
+
+size_t Chessboard::Board::rowCount()const
+{
+    return rows;
+}
+
+cv::Size Chessboard::Board::getSize()const
+{
+    return cv::Si
```

Comment:
Looks like color -> gray conversion is performed in [findChessboardCorners2](https://github.com/opencv/opencv/pull/12147/files#diff-8df6e9bc9d383d0c31ba9d371e491352R3263)

Reply:
[Code](https://github.com/opencv/opencv/pull/12147/files#diff-8df6e9bc9d383d0c31ba9d371e491352R3101):
```
    cv::Mat gray;
    switch(image.type())
    {
    case CV_8UC3:
        cv::cvtColor(image,gray,cv::COLOR_BGR2GRAY);
    case CV_8UC1:
        gray = image;
        break;
    default:
        CV_Error(Error::StsUnsupportedFormat,"unsupported color format");
    }
```